### ISSUE TITLE
Potential bug on Forward_tag short-circuiting?

### ISSUE THREAD

**Issue by gasche** [ocamllabs/ocaml-multicore#749]
## Context

When the minor GC sees a `Forward_tag` block, it can short-circuit the block by rewriting it to hold the value. (Forward_tag denotes a lazy value whose thunk has already been successfully forced; short-circuiting makes the memory representation more compact.)

Short-circuiting must be careful to not introduce confusions: it is only valid for values that cannot be mistaken with lazy thunks, and (if flat-float-array) it cannot introduce `float` values in the type of lazy values. To avoid confusions, short-circuiting is thus disabled on values of type `Forward_tag`, `Lazy_tag` and `Double_tag`. See the [minor_gc.c code](https://github.com/ocaml-multicore/ocaml-multicore/blob/4469f89/runtime/minor_gc.c#L363-L381):

## Potential bug

Multicore introduces a new tag for lazy thunks, `Obj.forcing_tag`, for values that are in the process of being forced (possibly concurrently). I believe that it is incorrect to shortcut `Forward_tag` on values of type `Obj.forcing_tag`, due to the following scenario:

```ocaml
let inner_thunk = lazy ...
let outer_thunk = lazy (some_allocating_code; inner_thunk)

let () = Lazy.force outer_thunk
  (* outer_thunk is now a Forward block to inner_thunk *)

let () = Domain.spawn (fun () ->
  ignore (Lazy.force inner_thunk)
  (* sets `inner_thunk` to Forcing_tag, then pause with a GC that short-circuits `outer_thunk` *))
let () = Domain.spawn (fun () ->
  ignore (Lazy.force outer_thunk)
  (* observes Forcing_tag as the tag of the short-circuited outer_thunk, and fails *) )
```

**Comment by gasche** [comment]
The same issue may also exist in `Lazy.from_val`:

```ocaml
let from_val (v : 'arg) =
  let t = Obj.tag (Obj.repr v) in
  if t = Obj.forward_tag || t = Obj.lazy_tag || t = Obj.double_tag then begin
    make_forward v
  end else begin
    (Obj.magic v : 'arg t)
  end
```

I believe that `|| t = Obj.forcing_tag` should also be added there for similar reasons.

**Comment by kayceesrk** [comment]
This is a good catch. Both of these cases need to be fixed.

**Comment by kayceesrk** [comment]
Fixed by #760.
